### PR TITLE
Add support for LongCat-Image

### PR DESCRIPTION
**Pull request by junqiangwu**
# What does this PR do?

This PR introduces the LongCat-Image Series into the diffusers library. The LongCat-Image Series comprises powerful and highly efficient 6B-parameter models, featuring both a state-of-the-art image generation model and a versatile image editing model. Designed to be friendly for consumer-grade hardware, this series demonstrates strong capabilities in high-fidelity image synthesis and precise, controllable image editing. The technical report and checkpoints for both models have been released.

Thanks for the support of @yiyixuxu @apolinario.

- GitHub: https://github.com/meituan-longcat/LongCat-Image
- LongCat-Image on Hugging Face: https://huggingface.co/meituan-longcat/LongCat-Image
- LongCat-Image-Edit on Hugging Face: https://huggingface.co/meituan-longcat/LongCat-Image-Edit

### CODE REVIEW

File: src/diffusers/models/transformers/transformer_longcat_image.py [hunk 1]
```diff
@@ -0,0 +1,568 @@
...
+                "Passing `txt_ids` 3d torch.Tensor is deprecated."
+                "Please remove the batch dimension and pass it as a 2d torch Tensor"
+            )
+            txt_ids = txt_ids[0]
+        if img_ids.ndim == 3:
+            logger.warning(
+                "Passing `img_ids` 3d torch.Tensor is deprecated."
+                "Please remove the batch dimension and pass it as a 2d torch Tensor"
+            )
+            img_ids = img_ids[0]
```
Comment (yiyixuxu) [inline src/diffusers/models/transformers/transformer_longcat_image.py]:
```suggestion
```
we don't need to deprecate for new pipeline :)

File: src/diffusers/models/transformers/transformer_longcat_image.py [hunk 2]
```diff
@@ -0,0 +1,568 @@
...
+            If `return_dict` is True, an [`~models.transformer_2d.Transformer2DModelOutput`] is returned, otherwise a
+            `tuple` where the first element is the sample tensor.
+        """
+        hidden_states = self.x_embedder(hidden_states)
+
+        timestep = timestep.to(hidden_states.dtype) * 1000
+        if guidance is not None:
+            guidance = guidance.to(hidden_states.dtype) * 1000
+        else:
+            guidance = None
```
Comment (yiyixuxu) [inline src/diffusers/models/transformers/transformer_longcat_image.py]:
```suggestion
```
it is not used I think?

File: src/diffusers/models/transformers/transformer_longcat_image.py [hunk 3]
```diff
@@ -0,0 +1,568 @@
...
+            )
+            cos_out.append(cos)
+            sin_out.append(sin)
+        freqs_cos = torch.cat(cos_out, dim=-1).to(ids.device)
+        freqs_sin = torch.cat(sin_out, dim=-1).to(ids.device)
+        return freqs_cos, freqs_sin
+
+
+
+class TimestepEmbeddings(nn.Module):
```
Comment (yiyixuxu) [inline src/diffusers/models/transformers/transformer_longcat_image.py]:
```suggestion
class LongCatImageTimestepEmbeddings(nn.Module):
```

File: src/diffusers/models/transformers/transformer_longcat_image.py [hunk 4]
```diff
@@ -0,0 +1,568 @@
...
+        axes_dims_rope: List[int] = [16, 56, 56],
+    ):
+        super().__init__()
+        self.out_channels = in_channels
+        self.inner_dim = num_attention_heads * attention_head_dim
+        self.pooled_projection_dim = pooled_projection_dim
+
+        self.pos_embed = LongCatImagePosEmbed(theta=10000, axes_dim=axes_dims_rope)
+
+        self.time_embed = TimestepEmbeddings(embedding_dim=self.inner_dim)
```
Comment (yiyixuxu) [inline src/diffusers/models/transformers/transformer_longcat_image.py]:
```suggestion
        self.time_embed = LongCatImageTimestepEmbeddings(embedding_dim=self.inner_dim)
```

File: src/diffusers/pipelines/longcat_image/pipeline_longcat_image.py [hunk 5]
```diff
@@ -0,0 +1,693 @@
...
+        processing larger images.
+        """
+        self.vae.enable_tiling()
+
+    def disable_vae_tiling(self):
+        r"""
+        Disable tiled VAE decoding. If `enable_vae_tiling` was previously enabled, this method will go back to
+        computing decoding in one step.
+        """
+        self.vae.disable_tiling()
```
Comment (yiyixuxu) [inline src/diffusers/pipelines/longcat_image/pipeline_longcat_image.py]:
```suggestion
```

File: src/diffusers/pipelines/longcat_image/pipeline_longcat_image.py [hunk 6]
```diff
@@ -0,0 +1,693 @@
...
+
+        _, seq_len, _ = prompt_embeds.shape
+
+        # duplicate text embeddings and attention mask for each generation per prompt, using mps friendly method
+        prompt_embeds = prompt_embeds.repeat(1, num_images_per_prompt, 1)
+        prompt_embeds = prompt_embeds.view(batch_size * num_images_per_prompt, seq_len, -1)
+
+        return prompt_embeds
+
+    @torch.inference_mode()
```
Comment (yiyixuxu) [inline src/diffusers/pipelines/longcat_image/pipeline_longcat_image.py]:
```suggestion
```

File: src/diffusers/pipelines/longcat_image/pipeline_longcat_image.py [hunk 7]
```diff
@@ -0,0 +1,693 @@
...
+        
+        self.prompt_template_encode_prefix = '<|im_start|>system\nAs an image captioning expert, generate a descriptive text prompt based on an image content, suitable for input to a text-to-image model.<|im_end|>\n<|im_start|>user\n'
+        self.prompt_template_encode_suffix = '<|im_end|>\n<|im_start|>assistant\n'
+        self.prompt_template_encode_start_idx = 36
+        self.prompt_template_encode_end_idx = 5
+        self.default_sample_size = 128
+        self.tokenizer_max_length = 512
+
+
+    @torch.inference_mode()
```
Comment (yiyixuxu) [inline src/diffusers/pipelines/longcat_image/pipeline_longcat_image.py]:
```suggestion
```

File: src/diffusers/pipelines/longcat_image/pipeline_longcat_image.py [hunk 8]
```diff
@@ -0,0 +1,693 @@
...
+        generated_ids_trimmed = [
+            out_ids[len(in_ids) :] for in_ids, out_ids in zip(inputs.input_ids, generated_ids)
+        ]
+        output_text = self.text_processor.batch_decode(
+            generated_ids_trimmed, skip_special_tokens=True, clean_up_tokenization_spaces=False
+        )[0]
+        rewrite_prompt= output_text
+        return rewrite_prompt
+    
+    def _encode_prompt( self, prompt, num_images_per_prompt ):
```
Comment (yiyixuxu) [inline src/diffusers/pipelines/longcat_image/pipeline_longcat_image.py]:
```suggestion
    def _encode_prompt( self, prompt ):
```
can we move the part of code that duplicate using `num_images_per_prompt` inside `encde_prompt` instead?

**Review by yiyixuxu — COMMENTED** [review]
thanks a lot for the PR!
looks really good to me, I left some very small feedbacks, I think we can merge this soon:)

we also need doc & tests but can work on that in a follow-up PR if you prefer. Let us know

File: src/diffusers/pipelines/longcat_image/pipeline_longcat_image.py (hunk 8 above)
Comment (junqiangwu) [reply src/diffusers/pipelines/longcat_image/pipeline_longcat_image.py]: Sure, we’ve already updated this part of the code.

File: src/diffusers/pipelines/longcat_image/pipeline_longcat_image.py [hunk 9]
```diff
@@ -0,0 +1,661 @@
...
+        ]
+        output_text = self.text_processor.batch_decode(
+            generated_ids_trimmed, skip_special_tokens=True, clean_up_tokenization_spaces=False
+        )[0]
+        rewrite_prompt= output_text
+        return rewrite_prompt
+    
+    def _encode_prompt( self, prompt ):
+        all_tokens = []
+        for clean_prompt_sub, matched in split_quotation(prompt[0]):
```
Comment (yiyixuxu) [inline src/diffusers/pipelines/longcat_image/pipeline_longcat_image.py]: are we only support single prompt? if so maybe can add a check in check_inputs

File: src/diffusers/pipelines/longcat_image/pipeline_longcat_image.py [hunk 10]
```diff
@@ -0,0 +1,661 @@
...
+                    tokens = self.tokenizer(sub_word, add_special_tokens=False)['input_ids']
+                    all_tokens.extend(tokens)
+            else:
+                tokens = self.tokenizer(clean_prompt_sub, add_special_tokens=False)['input_ids']
+                all_tokens.extend(tokens)
+        
+        if len(all_tokens) > self.tokenizer_max_length:
+            logger.warning(
+                "Your input was truncated because `max_sequence_length` is set to "
+                f" {self.tokenizer_max_length} input token nums : {len(len(all_tokens))}"
```
Comment (yiyixuxu) [inline src/diffusers/pipelines/longcat_image/pipeline_longcat_image.py]:
```suggestion
                f" {self.tokenizer_max_length} input token nums : {len(all_tokens)}"
```

File: src/diffusers/pipelines/longcat_image/pipeline_longcat_image.py [hunk 11]
```diff
@@ -0,0 +1,661 @@
...
+            text_processor=text_processor
+        )
+
+        self.vae_scale_factor = 2 ** (len(self.vae.config.block_out_channels) - 1) if getattr(self, "vae", None) else 8
+        self.image_processor = VaeImageProcessor(vae_scale_factor=self.vae_scale_factor * 2)
+        
+        self.prompt_template_encode_prefix = '<|im_start|>system\nAs an image captioning expert, generate a descriptive text prompt based on an image content, suitable for input to a text-to-image model.<|im_end|>\n<|im_start|>user\n'
+        self.prompt_template_encode_suffix = '<|im_end|>\n<|im_start|>assistant\n'
+        self.prompt_template_encode_start_idx = 36
+        self.prompt_template_encode_end_idx = 5
```
Comment (yiyixuxu) [inline src/diffusers/pipelines/longcat_image/pipeline_longcat_image.py]:
```suggestion
```
with the method we process promptsfor longcat, we can calculate this dynamically pretty easy, no?

File: src/diffusers/pipelines/longcat_image/pipeline_longcat_image.py [hunk 12]
```diff
@@ -0,0 +1,661 @@
...
+        input_ids = input_ids.unsqueeze(0).to(self.device)
+        attention_mask = attention_mask.unsqueeze(0).to(self.device)
+
+        text_output = self.text_encoder(
+            input_ids=input_ids,
+            attention_mask=attention_mask,
+            output_hidden_states=True
+        )
+        # [max_sequence_length, batch, hidden_size] -> [batch, max_sequence_length, hidden_size]
+        # clone to have a contiguous tensor
```
Comment (yiyixuxu) [inline src/diffusers/pipelines/longcat_image/pipeline_longcat_image.py:330]:
```suggestion
        # clone to have a contiguous tensor
        self.prompt_template_encode_start_idx = len(prefix_tokens)
        self.prompt_template_encode_end_idx = len(suffix_tokens)
```

File: src/diffusers/pipelines/longcat_image/pipeline_longcat_image.py [hunk 13]
```diff
@@ -0,0 +1,661 @@
...
+
+        Returns:
+            [`~pipelines.LongCatImagePipelineOutput`] or `tuple`: [`~pipelines.LongCatImagePipelineOutput`] if `return_dict`
+            is True, otherwise a `tuple`. When returning a tuple, the first element is a list with the generated
+            images.
+        """
+
+        height = height or self.default_sample_size * self.vae_scale_factor
+        width = width or self.default_sample_size * self.vae_scale_factor
+        if height % (self.vae_scale_factor * 2) != 0 or width % (self.vae_scale_factor * 2) != 0:
```
Comment (yiyixuxu) [inline src/diffusers/pipelines/longcat_image/pipeline_longcat_image.py]: can you add a check_inputs method? like this https://github.com/huggingface/diffusers/blob/main/src/diffusers/pipelines/qwenimage/pipeline_qwenimage.py#L564

if we only support single prompt, we should add a check there too

File: src/diffusers/pipelines/longcat_image/pipeline_longcat_image.py [hunk 14]
```diff
@@ -0,0 +1,661 @@
...
+        # 2. Define call parameters
+        if prompt is not None and isinstance(prompt, str):
+            batch_size = 1
+        elif prompt is not None and isinstance(prompt, list):
+            batch_size = len(prompt)
+        else:
+            batch_size = prompt_embeds.shape[0]
+        
+        device = self._execution_device
+        # from IPython import embed; embed();
```
Comment (yiyixuxu) [inline src/diffusers/pipelines/longcat_image/pipeline_longcat_image.py]:
```suggestion

```

File: src/diffusers/pipelines/longcat_image/pipeline_longcat_image_edit.py [hunk 15]
```diff
@@ -0,0 +1,730 @@
...
+        ...     prompt,
+        ...     num_inference_steps=50,
+        ...     guidance_scale=4.5,
+        ...     generator=torch.Generator("cpu").manual_seed(43),
+        ... ).images[0]
+        >>> image.save("longcat_image_edit.png")
+        ```
+"""
+
+def split_quotation(prompt, quote_pairs=None):
```
Comment (yiyixuxu) [inline src/diffusers/pipelines/longcat_image/pipeline_longcat_image_edit.py:72]: is this same as one in longcat pipeline? if so, can we add a `#Copied from` statement?

File: src/diffusers/pipelines/longcat_image/pipeline_longcat_image_edit.py [hunk 16]
```diff
@@ -0,0 +1,730 @@
...
+        if re.match(pattern, part):
+            if len(part):
+                result.append((part, True))
+        else:
+            if len(part):
+                result.append((part, False))
+    return result
+
+
+def prepare_pos_ids(
```
Comment (yiyixuxu) [inline src/diffusers/pipelines/longcat_image/pipeline_longcat_image_edit.py]: same here, we should add a `#Copied from` if it's same as in the other pipeline

File: src/diffusers/pipelines/longcat_image/pipeline_longcat_image_edit.py [hunk 17]
```diff
@@ -0,0 +1,730 @@
...
+            output_hidden_states=True
+        )
+        # [max_sequence_length, batch, hidden_size] -> [batch, max_sequence_length, hidden_size]
+        # clone to have a contiguous tensor
+        prompt_embeds = text_output.hidden_states[-1].detach()
+        prompt_embeds = prompt_embeds[:,self.prompt_template_encode_start_idx: -self.prompt_template_encode_end_idx ,:]
+
+        return prompt_embeds
+
+    @torch.inference_mode()
```
Comment (yiyixuxu) [inline src/diffusers/pipelines/longcat_image/pipeline_longcat_image_edit.py]:
```suggestion
```

File: src/diffusers/pipelines/longcat_image/pipeline_longcat_image_edit.py [hunk 18]
```diff
@@ -0,0 +1,730 @@
...
+        processing larger images.
+        """
+        self.vae.enable_tiling()
+
+    def disable_vae_tiling(self):
+        r"""
+        Disable tiled VAE decoding. If `enable_vae_tiling` was previously enabled, this method will go back to
+        computing decoding in one step.
+        """
+        self.vae.disable_tiling()
```
Comment (yiyixuxu) [inline src/diffusers/pipelines/longcat_image/pipeline_longcat_image_edit.py]:
```suggestion
```

File: src/diffusers/pipelines/longcat_image/pipeline_longcat_image_edit.py [hunk 19]
```diff
@@ -0,0 +1,730 @@
...
+
+        Returns:
+            [`~pipelines.LongCatImagePipelineOutput`] or `tuple`: [`~pipelines.LongCatImagePipelineOutput`] if `return_dict`
+            is True, otherwise a `tuple`. When returning a tuple, the first element is a list with the generated
+            images.
+        """
+
+        image_size = image.size
+        calculated_width, calculated_height = calculate_dimensions(1024 * 1024, image_size[0]*1.0/image_size[1])
+
```
Comment (yiyixuxu) [inline src/diffusers/pipelines/longcat_image/pipeline_longcat_image_edit.py]: can we add a `check_inputs` method in this pipeine too?

**Review by yiyixuxu — COMMENTED** [review]
thanks! I left a few more comments

**Comment by yiyixuxu** [comment]
I think you need to add the doc page into the index here https://github.com/huggingface/diffusers/blob/main/docs/source/en/_toctree.yml

also can you run `make style` and `make fix-copies`?

File: src/diffusers/pipelines/longcat_image/pipeline_longcat_image.py (hunk 9 above)
Comment (junqiangwu) [reply src/diffusers/pipelines/longcat_image/pipeline_longcat_image.py]: We’ve updated the code so that the text-to-image pipeline now accepts multiple prompts, while the edit pipeline still takes a single prompt. Both behaviors are enforced via checks in check_inputs.

**Comment by junqiangwu** [comment]
Thanks for the feedback — we’ve updated the code accordingly and run make style and make fix-copies in this commit. Please let us know if you spot anything else that should be addressed. @yiyixuxu

**Comment by yiyixuxu** [comment]
@bot /style

**Comment by junqiangwu** [comment]
My apologies. During further testing with additional examples, I identified an issue and have already implemented a fix. I’m not completely sure whether this change might affect your merge.

**Comment by ovi054** [comment]
load_lora_weights(), unload_lora_weights these doesn't work with LongCatImagePipeline. @junqiangwu
image
